### PR TITLE
output location feature added.

### PR DESCRIPTION
**Pull request by d-eMason**
### Individual files:
- encrypt with given file name:
- decrypt with given file name:

### Folders:
-  encrypt with given output folder location:
- decrypt with given output folder location:

_refer to the test_output_location.md files for test cases._

### CODE REVIEW

File: efc [hunk 1]
```diff
@@ -145,40 +147,92 @@ user_input() {
 }
 
 process_all_files_in_dir() {
+
+    if [ -z "${2}" ]; then
+        echo "output location is not provided, using the same location of the source file."
```
Comment (compilable) [inline efc:152]: can follow info/warn logs format?

File: efc [hunk 2]
```diff
@@ -145,40 +147,92 @@ user_input() {
...
+        OUTPUT_DIR="$1"
+    else
+        # create the folder if not exists
+        mkdir -p "$2"
+        if test -d "${2}"; then
+            OUTPUT_DIR="$2"
+            echo "output location is provided, using the location: $OUTPUT_DIR"
```
Comment (compilable) [inline efc:159]: can follow info/warn logs format?

File: efc_lib [hunk 3]
```diff
@@ -53,10 +50,10 @@ encrypt() {
     echo -e ' \t' "encrypting the file: $1"
 
     if [ -z "${4}" ]; then
-        # encrypt to the same file name.gpg
+        echo "encrypt to the same file name.gpg"
```
Comment (compilable) [inline efc_lib:53]: excessive logs, since this is a lib, can we leave this log as a commit ?

File: efc_lib [hunk 4]
```diff
@@ -53,10 +50,10 @@ encrypt() {
...
         gpg --quiet --yes --batch --passphrase "$2" -c "$1"
     else
-        # encrypt to a given file name
+        echo "encrypt to a given file name $4"
```
Comment (compilable) [inline efc_lib:56]: excessive logs, since this is a lib, can we leave this log as a commit ?

File: efc [hunk 5]
```diff
@@ -145,40 +147,92 @@ user_input() {
...
     find "$1" -print |
         while read file; do
             echo "FILE :: $file"
 
             if [[ -d $file ]]; then
+
                 echo "Insdie : DIR  $file"
+                # create the folder in output location if OUTPUT_DIR is provided
+                relative_path=$(realpath --relative-to="$1" "$file")
+                echo "xxxxxxxxxxxx $OUTPUT_DIR/$relative_path"
```
Comment (compilable) [inline efc:174]: please remove the echo.

File: efc [hunk 6]
```diff
@@ -145,40 +147,92 @@ user_input() {
...
+                mkdir -p "$OUTPUT_DIR/$relative_path"
             else
 
+                # generate the file name if OUTPUT_DIR is provided
+                src_name=$(basename "$file")
```
Comment (compilable) [inline efc:179]: are we using the src_name ?

File: efc [hunk 7]
```diff
@@ -145,40 +147,92 @@ user_input() {
...
+                relative_path=$(realpath --relative-to="$1" "$src_path")
+                # remove . path from relative path
+                #relative_path="${relative_path//./""}"
+
                 if [[ $isEnrypt == 'e' ]]; then
-                    #encrypt "$file" "$password" $isDelete
                     echo "ENC" $file
-                    encrypt "$file" "$password" $isDelete
+                    OUTPUT_FILE="$OUTPUT_DIR/$relative_path/$src_name.$ENCRYPTED_FILE_EXT"
+                    echo "OUTD :: $OUTPUT_FILE"
```
Comment (compilable) [inline efc:188]: can we follow info/warn logs format?

File: efc [hunk 8]
```diff
@@ -145,40 +147,92 @@ user_input() {
...
                     ((counter++))
                 fi
 
                 if [[ $isEnrypt == 'd' ]]; then
 
                     echo "DEC" $file
-                    decrypt "$file" "$password" $isDelete
+                    src_name="${src_name%.*}"
+                    OUTPUT_FILE="$OUTPUT_DIR/$relative_path/$src_name"
+                    echo "OUTD :: $OUTPUT_FILE"
```
Comment (compilable) [inline efc:199]: can we follow info/warn logs format?

File: efc (hunk 1 above)
Comment (d-eMason) [reply efc:152]: echo "INFO:  added.

File: efc (hunk 2 above)
Comment (d-eMason) [reply efc:159]: echo "INFO:  added.

**Pull request merged**